### PR TITLE
Log around lock operations

### LINKED ISSUE #11622: [lock] add logging to ILock operations
Currently `ILock` operations don't appear to be logged at any level.  It would be very helpful if I could enable diagnostic log output, with the log line containing the name of the ILock, the operation (`create`, `lock`, `unlock`, `destroy`) and the usual timestamp, thread ID etc.

This would not only help diagnose locking problems (order of operations etc), but it would also help expose locks that are not being destroyed after use.

### LINKED ISSUE #11683: [scheduled-executor] ScheduledExecutorServiceSlowTest.schedule_thenDisposeLeakTest
```
org.junit.runners.model.TestTimedOutException: test timed out after 600000 milliseconds
	at java.lang.System.identityHashCode(Native Method)
	at java.io.ObjectOutputStream$HandleTable.hash(ObjectOutputStream.java:2360)
	at java.io.ObjectOutputStream$HandleTable.insert(ObjectOutputStream.java:2323)
	at java.io.ObjectOutputStream$HandleTable.assign(ObjectOutputStream.java:2281)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1428)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.write(JavaDefaultSerializers.java:242)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.write(StreamSerializerAdapter.java:43)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.writeObject(AbstractSerializationService.java:250)
	at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataOutput.writeObject(ByteArrayObjectDataOutput.java:370)
	at com.hazelcast.scheduledexecutor.impl.TaskDefinition.writeData(TaskDefinition.java:119)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.write(DataSerializableSerializer.java:201)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.write(DataSerializableSerializer.java:50)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.write(StreamSerializerAdapter.java:43)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.writeObject(AbstractSerializationService.java:250)
	at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataOutput.writeObject(ByteArrayObjectDataOutput.java:370)
	at com.hazelcast.scheduledexecutor.impl.operations.ScheduleTaskOperation.writeInternal(ScheduleTaskOperation.java:60)
	at com.hazelcast.spi.Operation.writeData(Operation.java:565)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.write(DataSerializableSerializer.java:201)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.write(DataSerializableSerializer.java:50)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.write(StreamSerializerAdapter.java:43)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:152)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:128)
	at com.hazelcast.spi.impl.operationservice.impl.OutboundOperationHandler.send(OutboundOperationHandler.java:51)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvokeRemote(Invocation.java:574)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:552)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:513)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:207)
	at com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl.invokeOnPartition(OperationServiceImpl.java:318)
	at com.hazelcast.spi.AbstractDistributedObject.invokeOnPartition(AbstractDistributedObject.java:72)
	at com.hazelcast.scheduledexecutor.impl.ScheduledExecutorServiceProxy.submitOnPartitionSync(ScheduledExecutorServiceProxy.java:428)
	at com.hazelcast.scheduledexecutor.impl.ScheduledExecutorServiceProxy.schedule(ScheduledExecutorServiceProxy.java:110)
	at com.hazelcast.scheduledexecutor.ScheduledExecutorServiceSlowTest.schedule_thenDisposeLeakTest(ScheduledExecutorServiceSlowTest.java:423)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast%20Maintenance/job/Hazelcast-3.maintenance-nightly/com.hazelcast$hazelcast/1384/testReport/junit/com.hazelcast.scheduledexecutor/ScheduledExecutorServiceSlowTest/schedule_thenDisposeLeakTest/

### PR DESCRIPTION
**Pull request by blazember**
Log at the finest level when locks are created, acquired, released or destroyed. Log messages added both before and after the operation with contextual information (success, requester, owner) where applicable.

Fixes #11622

### CODE REVIEW

**Comment by mmedenjak** [comment]
run-lab-run

**Comment by blazember** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/UnlockOperation.java [hunk 1]
```diff
@@ -65,19 +66,44 @@ public void run() throws Exception {
...
         LockStoreImpl lockStore = getLockStore();
-        response = lockStore.forceUnlock(key);
+        boolean unlocked = lockStore.forceUnlock(key);
+        this.response = unlocked;
+
+        if (logger.isFinestEnabled()) {
+            if (unlocked) {
+                logger.finest("Released lock " + namespace.getObjectName());
+            } else {
+                logger.finest("Could not release lock " + namespace.getObjectName()
```
Comment (kwart) [inline hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/UnlockOperation.java]: Could you double check this? It seems the `false` is returned if no such lock exist.

The `LockStoreImpl` contains:
```
    public boolean forceUnlock(Data key) {
        LockResourceImpl lock = locks.get(key);
        if (lock == null) {
            return false;
        }
...
```

File: hazelcast/src/main/java/com/hazelcast/concurrent/lock/LockServiceImpl.java [hunk 2]
```diff
@@ -328,11 +332,21 @@ private void clearLockStoresHavingLesserBackupCountThan(int partitionId, int thr
 
     @Override
     public DistributedObject createDistributedObject(String objectId) {
-        return new LockProxy(nodeEngine, this, objectId);
+        LockProxy lockProxy = new LockProxy(nodeEngine, this, objectId);
+
+        if (logger.isFinestEnabled()) {
+            logger.finest("Created lock " + objectId);
```
Comment (kwart) [inline hazelcast/src/main/java/com/hazelcast/concurrent/lock/LockServiceImpl.java]: IMO, we should talk only about a *"LockProxy"* here and not the *"lock"*.

If I remember correctly, the proxy can exist without the real object in the cluster.

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/concurrent/lock/LockServiceImpl.java]: That's right, the proxy merely sends operations and is agnostic as to the existence of the actual lock container

File: hazelcast/src/main/java/com/hazelcast/concurrent/lock/LockStore.java [hunk 3]
```diff
@@ -157,7 +157,8 @@
      * waiters/signals/expired operations.
      *
      * @param dataKey the lock key
-     * @return if the key was unlocked or is already unlocked
+     * @return if the key was unlocked or is already unlocked but still in the store. Returns false only if the key is
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/concurrent/lock/LockStore.java]: Can you rephrase this a bit to something like `@return {@code false} if there is no lock for the given {@code key} and {@code true} otherwise`

File: hazelcast/src/main/java/com/hazelcast/concurrent/lock/LockServiceImpl.java [hunk 4]
```diff
@@ -338,11 +342,21 @@ private void clearLockStoresHavingLesserBackupCountThan(int partitionId, int thr
...
+        if (logger.isFinestEnabled()) {
+            logger.finest("Created proxy for lock " + objectId);
+        }
+
+        return lockProxy;
     }
 
     @Override
     public void destroyDistributedObject(String objectId) {
+        if (logger.isFinestEnabled()) {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/concurrent/lock/LockServiceImpl.java]: I would prefer removing the proxy creation logging. Proxies are irrelevant. The most important part is what is happening on the partition owner.

File: hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/LockOperation.java [hunk 5]
```diff
@@ -44,7 +45,25 @@ public LockOperation(ObjectNamespace namespace, Data key, long threadId, long le
 
     @Override
     public void run() throws Exception {
-        response = getLockStore().lock(key, getCallerUuid(), threadId, getReferenceCallId(), leaseTime);
+        ILogger logger = getLogger();
+
+        if (logger.isFinestEnabled()) {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/LockOperation.java]: I think it is enough to log the end result and skip logging the intent (`Acquiring lock...`). Same for the `UnlockOperation`.

File: hazelcast/src/main/java/com/hazelcast/concurrent/lock/LockServiceImpl.java [hunk 6]
```diff
@@ -88,6 +90,8 @@ public LockServiceImpl(NodeEngine nodeEngine) {
         }
 
         maxLeaseTimeInMillis = getMaxLeaseTimeInMillis(nodeEngine.getProperties());
+
+        this.logger = nodeEngine.getLogger(getClass());
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/concurrent/lock/LockServiceImpl.java]: You can remove this logger if it's no longer used.
